thirdparty/DiskANN/src/aux_utils.cpp

```diff
@@ -602,7 +606,279 @@ namespace diskann {
+                                 const std::vector<unsigned> &cached_nodes_list,
+                                 std::vector<std::vector<unsigned>> &graph) {
+    std::ofstream cache_writer(cache_file.c_str(), std::ios::binary);
+    std::ifstream base_reader(base_file, std::ios::binary);
+
+    uint32_t npts_u32, dim_u32;
+    base_reader.read((char *) &npts_u32, sizeof(uint32_t));
+    base_reader.read((char *) &dim_u32, sizeof(uint32_t));
+
+    uint32_t R_u32 = R;
+    uint32_t aligned_dim = ROUND_UP(dim_u32, 8);
+    auto     raw_data_buf = std::vector<T>(aligned_dim);
+    auto     neighbors_buf = std::vector<unsigned>(R + 1);
+    cache_writer.write((char *) &npts_u32, sizeof(uint32_t));
+    cache_writer.write((char *) &R_u32, sizeof(uint32_t));
+    cache_writer.write((char *) &aligned_dim, sizeof(uint32_t));
+
+    for (size_t cache_i = 0; cache_i < cached_nodes_list.size(); cache_i++) {
+      auto vec_id = cached_nodes_list[cache_i];
+      // write id
+      cache_writer.write((char *) &vec_id, sizeof(unsigned));
+      // write neighbors
+      uint32_t neighbors_num = graph[vec_id].size();
+      neighbors_buf[0] = neighbors_num;
+      memcpy(neighbors_buf.data() + 1, graph[vec_id].data(),
+             sizeof(unsigned) * graph[vec_id].size());
+      cache_writer.write((char *) neighbors_buf.data(),
+                         sizeof(unsigned) * (R + 1));
+      // write vector
+      uint64_t data_offest =
+          sizeof(uint32_t) * 2 + vec_id * dim_u32 * sizeof(T);
+      base_reader.seekg(data_offest, base_reader.beg);
+      base_reader.read((char *) raw_data_buf.data(), sizeof(T) * dim_u32);
+      cache_writer.write((char *) raw_data_buf.data(), sizeof(T) * aligned_dim);
+    }
+    base_reader.close();
+    cache_writer.close();
+  }
+
+  template<typename T>
+  void load_cached_nodes_from_file(const unsigned aligned_dim, const unsigned R,
+                                   unsigned              &cached_nodes_num,
+                                   const std::string     &cache_file,
+                                   std::vector<uint32_t> &node_list,
+                                   unsigned *&neighbors_buf, T *&vec_buf) {
+    std::ifstream cache_reader(cache_file, std::ios::binary);
+    uint32_t      npts_u32, r_u32, aligned_dim_u32;
+    cache_reader.read((char *) &npts_u32, sizeof(uint32_t));
+    cache_reader.read((char *) &r_u32, sizeof(uint32_t));
+    cache_reader.read((char *) &aligned_dim_u32, sizeof(uint32_t));
+    if (r_u32 != R && aligned_dim != aligned_dim_u32)
+      throw ANNException(
+          "Mismatch in R and aligned dim between cached_nodes_file and index",
+          -1, __FUNCSIG__, __FILE__, __LINE__);
+
+    cached_nodes_num = npts_u32;
+
+    uint32_t vec_size = aligned_dim_u32;
+    uint32_t neighbrs_size = (r_u32 + 1);
+    node_list = std::vector<unsigned>(cached_nodes_num);
+
+    neighbors_buf = new unsigned[cached_nodes_num * neighbrs_size];
```

Comment:
When is this deleted? Will this pointer already be allocated?

Reply:
This modification has now been removed.